File: flureadium/ios/flureadium/Sources/flureadium/ReadiumReaderChannel.swift
import Foundation
import Flutter
import ReadiumShared

class ReadiumReaderChannel: FlutterMethodChannel {
  // Compiles fine without this init, but then mysteriously crashes later with EXC_BAD_ACCESS when calling any member function later.
  init(name: String, binaryMessenger messenger: FlutterBinaryMessenger) {
    super.init(
      name: name, binaryMessenger: messenger, codec: FlutterStandardMethodCodec.sharedInstance(),
      taskQueue: nil)
  }

  func onPageChanged(locator: Locator) {
    invokeMethod("onPageChanged", arguments: locator.jsonString as String?)
  }

  func onExternalLinkActivated(url: URL) {
    invokeMethod("onExternalLinkActivated", arguments: url.absoluteString as String?)
  }

  /// Reports a tap on content, in points relative to the navigator's view.
  /// Dart decodes the pair as an `Offset`.
  func onTap(position: CGPoint) {
    invokeMethod("onTap", arguments: ["x": Double(position.x), "y": Double(position.y)])
  }
}

File: flureadium/android/src/main/kotlin/dev/mulev/flureadium/ReadiumReaderChannel.kt
package dev.mulev.flureadium

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.util.AbsoluteUrl
import org.readium.r2.shared.util.toUri

internal class ReadiumReaderChannel(messenger: BinaryMessenger, name: String) :
    MethodChannel(messenger, name) {
    fun onPageChanged(locator: Locator?) =
        invokeMethod("onPageChanged", locator?.toJSON().toString())

    fun onExternalLinkActivated(url: AbsoluteUrl) =
        invokeMethod("onExternalLinkActivated", url.toString())

    fun onTap(x: Double, y: Double) =
        invokeMethod("onTap", mapOf("x" to x, "y" to y))
}
